node/src/components/block_synchronizer/global_state_synchronizer.rs

```diff
@@ -25,28 +25,52 @@ use crate::{
     },
     reactor,
     types::{BlockHash, NodeId, TrieOrChunk},
+    utils::DisplayIter,
     NodeRng,
 };
 
 const COMPONENT_NAME: &str = "global_state_synchronizer";
 
 #[derive(Debug, Clone, Error)]
 pub(crate) enum Error {
-    #[error(transparent)]
-    TrieAccumulator(TrieAccumulatorError),
-    #[error("ContractRuntime failed to put a trie into global state: {0}")]
-    PutTrie(engine_state::Error),
+    #[error("trie accumulator encountered an error while fetching a trie; unreliable peers {}", DisplayIter::new(.0))]
+    TrieAccumulator(Vec<NodeId>),
+    #[error("ContractRuntime failed to put a trie into global state: {0}; unreliable peers {}", DisplayIter::new(.1))]
+    PutTrie(engine_state::Error, Vec<NodeId>),
     #[error("no peers available to ask for a trie: {0}")]
     NoPeersAvailable(Digest),
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct Response {
+    hash: Digest,
+    unreliable_peers: Vec<NodeId>,
+}
+
+impl Response {
+    pub(crate) fn new(hash: Digest, unreliable_peers: Vec<NodeId>) -> Self {
+        Self {
+            hash,
+            unreliable_peers,
+        }
+    }
+
+    pub(crate) fn hash(&self) -> &Digest {
+        &self.hash
+    }
+
+    pub(crate) fn unreliable_peers(self) -> Vec<NodeId> {
+        self.unreliable_peers
+    }
+}
+
```

Comment:
you could do a two state enum for Success / Failure, and flatten the Error as a field on the failure variant; this would allow places where Result<Response, Error> to be simplified to just return the enum variant

Reply:
Yes, we could to that though I'm more partial to having separate types that are clearly delimited. What I think we can do is have a trait implemented for both the `Response` and `Error` that can extract the unreliable peers which would make the code looking at the response type potentially more nicer. WDYT?